decisions/0004-sync-routes-from-k8s-to-cc.md

```diff
@@ -0,0 +1,113 @@
+
+Kubernetes is also implemented via a series of "bulk sync loops" that are more commonly known as controllers, but there are some key differences in how these convergence loops work.
+1. k8s has a ["watch"](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) API so clients clients can subscribe to incremental changes in desired and actual state
+1. [k8s.io/client-go/cache](https://godoc.org/k8s.io/client-go/tools/cache) provides a mechanism for clients to build and maintain local caches of kubernetes state
+1. [kubebuilder](https://book.kubebuilder.io/) Provides, amongst other things, an SDK for building reconciliation loops on top of watches and client side caching
+
+### Problems and Solutions
+
+To illustrate the necessity of these sync loops, it helps to think about how they improve resiliency in 2 cases of temporary network partition from the perspective of the CF API: 
+1. Problem: Updates to CCDB succeed ✅, but updates to the Route CRD fail ❌
+1. Problem: Updates to the Route CRD succeed ✅, but updates to CCDB fail ❌
+
+To handle these 2 failure modes, several approaches can be taken. 
+
+##### Mimic historical sync loops:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, but allow the CCDB transaction to commit. A bulk sync loop will detect that the Route CRD is "missing" and create it later.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution A: Raise the error to the user, and don't do the Route CRD write until CCDB changes are committed.
+   1. Solution B: A bulk sync loop will detect that the Route CRD is "extraneous" and delete it later.
+
+##### Forgo a bulk sync and attempt to bootstrap consistency with k8s off CCDB transactions:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Roll back changes to the Route CRD... (but what if that still fails due to network partition?!)
+
+##### Sync from k8s to CCDB
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Raise the error to the user, but leave the Route CRD. The "bulk sync loop" will not consider the route "reconciled" until it is correct in CCDB.
+
+### Viable Long Term Implementations & Tradeoffs
+
+##### CCDB->k8s Sync in Ruby
+1. Mimics ProcessesSync for efficient access to CCDB
+1. Run it in an independent process from clock_global to avoid BOSH entanglement
+1. Unclear what scale limitations we might encounter if we regularly download
+   all k8s Routes.
+
+##### CCDB->k8s Sync in Go
+1. Probably a new, independent go program
+1. Can take advantage of go-client cache & reflector to avoid repeatedly
+   downloading all k8s state.
+   1. It's unclear, however, that kubebuilder could help
+   1. What does it mean to "reconcile" a change to a single Route CRD if CCDB is authoritative?
```

Comment:
I think you'd need to do something kind of hacky like what we did with our [metacontroller](https://metacontroller.app/) implementation. Have a singleton parent resource that represents the "sync" itself that the controller reconciles and have it automatically requeued on some interval.

What we had with [cfroutesync](https://github.com/cloudfoundry/cf-k8s-networking/tree/v0.1.0/cfroutesync) actually worked pretty well to be honest. It was just too slow to be the only means of route propagation.

Reply:
> a singleton parent resource that represents the "sync" itself

this might be the singleton missing piece that'd help us understand how we might do a CF->k8s sync with kubebuilder.

Reply:
Kubebuilder's controller-runtime has a concept of [GenericEvents](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.0/pkg/event?tab=doc#GenericEvent). This allows you to define your own sources that reconcilers can subscribe to. So you can still fit into the Kubebuilder model of building reconcilers that act on specific events occurring, they are just triggered on a custom source that triggers events whenever you need them to, whether that is just a polling event or something more sophisticated.